### PR TITLE
feat: Implement fat container for appsmith application

### PR DESCRIPTION
**Pull request by geekup-legodevops**
## Description

- Implement fat container for Appsmith ( include all service: redis, mongodb, rts, backend, editor app)
- Util command in fat container with 2 command export_db & import_db to export & restore database in fat container
- Fat container run without external docker.env configuration file

## Type of change
- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: deploy/fat_container/README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    image: containrrr/watchtower
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock
+    # Update check interval in seconds.
+    command: --interval 300 --label-enable --cleanup
+    networks:
+      - appsmith
+networks:
+  appsmith:
+    driver: bridge
```
Comment (sharat87) [inline deploy/fat_container/README.md:48]: Do we need the explicit network configuration here? Considering these two containers don't actually talk to each other at all?

```suggestion

  auto_update:
    image: containrrr/watchtower
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    # Update check interval in seconds.
    command: --interval 300 --label-enable --cleanup
```

File: deploy/fat_container/README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+To host Appsmith on a custom domain, you can contact your domain registrar and update your DNS records. Most domain registrars have documentation on how you can do this yourself.
+
+* [GoDaddy](https://in.godaddy.com/help/create-a-subdomain-4080)
+* [Amazon Route 53](https://aws.amazon.com/premiumsupport/knowledge-center/create-subdomain-route-53/)
+* [Digital Ocean](https://www.digitalocean.com/docs/networking/dns/how-to/add-subdomain/)
+* [NameCheap](https://www.namecheap.com/support/knowledgebase/article.aspx/9776/2237/how-to-create-a-subdomain-for-my-domain)
+* [Domain.com](https://www.domain.com/help/article/domain-management-how-to-update-subdomains)
+
+Then, you can update your custom domain from the application's configuration UI
+## 4. Utils Script
```
Comment (sharat87) [inline deploy/fat_container/README.md:72]: How about?

```suggestion
## 4. Instance Management Utilities
```

File: deploy/fat_container/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    image: containrrr/watchtower
+    volumes:
+      - /var/run/docker.sock:/var/run/docker.sock
+    # Update check interval in seconds.
+    command: --interval 300 --label-enable --cleanup
+    networks:
+      - appsmith
+networks:
+  appsmith:
+    driver: bridge
```
Comment (sharat87) [inline deploy/fat_container/docker-compose.yml:27]: Same as from README, if we don't need the explicit network configuration, I think we can remove it:

```suggestion

  auto_update:
    image: containrrr/watchtower
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    # Update check interval in seconds.
    command: --interval 300 --label-enable --cleanup

```

File: deploy/fat_container/README.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+Then, simply run following command to import data to internal database
+```
+docker exec appsmith_fat import_db
+```
+### 4.3 Supervisord UI
+To manage the application's processes using supervisord, you can use the supervisord UI
+
+You can use the browser to access port `9001` of the host (`http://localhost:9001` if you run the container on your local machine)
+<p>
+  <img src="./images/appsmith_supervisord_ui.png" width="80%">
```
Comment (sharat87) [inline deploy/fat_container/README.md:106]: Lets use an absolute URL to the image here instead of a relative path. The reason I say this is because GitHub will render this README at the URL `/appsmithorg/appsmith/deploy/fat_container`, as well as on `/appsmithorg/appsmith/deploy/fac_container/README.md`. But the way this relative link is resolved will be different in these two cases. In the first, it will resolve to `/appsmithorg/appsmith/deploy/images/appsmith_supervisord_ui.png` and in the second, to `/appsmithorg/appsmith/deploy/fat_container/images/appsmith_supervisord_ui.png`.

So its always best to use absolute links (not just for images) in README files on GitHub.

```suggestion
  <img src="https://raw.githubusercontent.com/appsmithorg/appsmith/release/deploy/fat_container/images/appsmith_supervisord_ui.png" width="80%">
```

File: deploy/fat_container/entrypoint.sh [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	AUTO_GEN_MONGO_PASSWORD=$(tr -dc A-Za-z0-9 </dev/urandom | head -c 13 ; echo '')
+	AUTO_GEN_ENCRYPTION_PASSWORD=$(tr -dc A-Za-z0-9 </dev/urandom | head -c 13 ; echo '')
+	AUTO_GEN_ENCRYPTION_SALT=$(tr -dc A-Za-z0-9 </dev/urandom | head -c 13 ; echo '')
+	bash "/opt/appsmith/templates/docker.env.sh" "$AUTO_GEN_MONGO_PASSWORD" "$AUTO_GEN_ENCRYPTION_PASSWORD" "$AUTO_GEN_ENCRYPTION_SALT" > "/opt/appsmith/configuration/docker.env"
+fi
+
+echo 'Load environment configuration'
+set -o allexport
+. /opt/appsmith/configuration/docker.env
+set +o allexport
```
Comment (sharat87) [inline deploy/fat_container/entrypoint.sh:155]: Should we check the existence of this file?

```suggestion
if [[ -f /opt/appsmith/configuration/docker.env ]]; then
  echo 'Load environment configuration'
  set -o allexport
  . /opt/appsmith/configuration/docker.env
  set +o allexport
fi
```

File: deploy/fat_container/entrypoint.sh [hunk 6]
```diff
@@ -0,0 +1,190 @@
+#!/usr/bin/env bash
```
Comment (sharat87) [inline deploy/fat_container/entrypoint.sh:1]: The indentation in this file seems inconsistent in several places, making it hard to follow the logic. Can we fix to using one indentation style (2-space or 4-space) and reformat this file please?

File: deploy/fat_container/scripts/renew-certificate.sh [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+	rsa_key_size=4096
+	
+	certbot certonly --webroot --webroot-path="$data_path/certbot" \
+		--register-unsafely-without-email \
+		--domains $domain \
+		--rsa-key-size $rsa_key_size \
+		--agree-tos \
+		--force-renewal
+	supervisorctl restart editor
+fi
```
Comment (sharat87) [inline deploy/fat_container/scripts/renew-certificate.sh:21]: Let's add a little note when custom domain is not configured, since this script effectively won't do anything in that case. When we have more docs on this topic, we can even share link to the relevant page in the output here.

```suggestion
else
  echo 'Custom domain not configured. Cannot enable SSL without a custom domain.' >&2
fi

```

File: deploy/fat_container/scripts/renew-certificate.sh [hunk 8]
```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
+
+set -e
+
+set -o allexport
+. /opt/appsmith/docker.env
+set +o allexport
+if [[ -n $CUSTOM_DOMAIN ]]; then
```
Comment (sharat87) [inline deploy/fat_container/scripts/renew-certificate.sh:8]: Minor check.

```suggestion
if [[ -f /opt/appsmith/docker.env ]]; then
  set -o allexport
  . /opt/appsmith/docker.env
  set +o allexport
fi

if [[ -n $CUSTOM_DOMAIN ]]; then
```

File: deploy/fat_container/templates/docker.env.sh [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+# Email server feature toggles
+# true | false values
+APPSMITH_MAIL_SMTP_AUTH=
+APPSMITH_MAIL_SMTP_TLS_ENABLED=
+
+# Disable all telemetry
+# Note: This only takes effect in self-hosted scenarios. 
+# Please visit: https://docs.appsmith.com/v/v1.2.1/setup/telemetry to read more about anonymized data collected by Appsmith
+# APPSMITH_DISABLE_TELEMETRY=false
```
Comment (sharat87) [inline deploy/fat_container/templates/docker.env.sh:60]: Let's uncomment env variables, where the value is a valid value (instead of an empty string). We've seen in the past where folks would change the value of an env variable, but forget to uncomment, and then come complaining on Discord that it doesn't work. Since the value `false` here is a valid value here, let's uncomment this proactively.

```suggestion
APPSMITH_DISABLE_TELEMETRY=false
```

File: deploy/fat_container/templates/mongo-init.js.sh [hunk 10]
```diff
@@ -0,0 +1,1539 @@
...
+            user: "$MONGO_ROOT_USER",
+            pwd: "$MONGO_ROOT_PASSWORD",
+            roles: [{
+                role: "root",
+                db: "admin"
+            }, "readWrite"]
+        }
+    ),
+
+    db.group.insert({
```
Comment (sharat87) [inline deploy/fat_container/templates/mongo-init.js.sh:27]: @mohanarpit, can we take this opportunity to remove the `group` and `propertyPane` inserts from this file? 😄

File: deploy/fat_container/templates/nginx_app.conf.sh [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+    client_max_body_size 100m;
+
+    gzip on;
+
+    root /opt/appsmith/editor;
+    index index.html index.htm;
+
+    location /.well-known/acme-challenge/ {
+        # root /var/www/certbot;
+		root /opt/appsmith/data/certificate/certbot;
```
Comment (sharat87) [inline deploy/fat_container/templates/nginx_app.conf.sh:26]: If the commented line is not required, or is unused, can we remove it? Let's try to avoid commented code where possible.

```suggestion
		root /opt/appsmith/data/certificate/certbot;
```

File: deploy/fat_container/templates/supervisord/application_process/cron.conf [hunk 12]
```diff
@@ -0,0 +1,12 @@
+[program:cron]
```
Comment (sharat87) [inline deploy/fat_container/templates/supervisord/application_process/cron.conf:1]: QQ, are we running our own cron daemon? Doesn't Debian come with one by default?

File: deploy/fat_container/utils/package.json [hunk 13]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "appsmith_utils",
+  "version": "1.0.0",
+  "description": "appsmith utils tool",
+  "main": "bin/index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC",
```
Comment (sharat87) [inline deploy/fat_container/utils/package.json:10]: Small adjustments.

```suggestion
  "license": "Apache-2.0",
  "private": true,
  "repository": {
    "type": "git",
    "url": "https://github.com/appsmithorg/appsmith.git",
    "directory": "deploy/fat_container"
  },
```

File: deploy/fat_container/utils/bin/import_db.js [hunk 14]
```diff
@@ -0,0 +1,47 @@
+// Init function export mongodb
```
Comment (sharat87) [inline deploy/fat_container/utils/bin/import_db.js:1]: Can we reformat this file? The indentation isn't consistent and it makes it hard to follow the logic and identify where a particular brace is closing etc.

File: deploy/fat_container/utils/bin/import_db.js [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        shell.exec('/usr/bin/supervisord')
+      }
+    })
+
+
+		shell.echo('stop backend & rts application before import database')
+		stop_application()
+		import_database()
+		shell.echo('start backend & rts application after import database')
+		start_application()
```
Comment (sharat87) [inline deploy/fat_container/utils/bin/import_db.js:38]: If there is an exception in `import_database`, that means the control flow will jump into the `catch` block, and this line for `start_application` will never be called right? Can we instead add a `finally` block and call `start_application()` inside the `finally` block? That way, the application will restart even if the import process failed or succeeded.

File: deploy/fat_container/utils/bin/export_db.js [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        shell.echo('application is not running, starting supervisord')
+        shell.exec('/usr/bin/supervisord')
+      }
+    })
+
+    shell.echo('stop backend & rts application before export database')
+    stop_application()
+    export_database()
+    shell.echo('start backend & rts application after export database')
+    start_application()
```
Comment (sharat87) [inline deploy/fat_container/utils/bin/export_db.js:42]: Same as my comment for import_db script, can we call `start_application()` from a `finally` block?

**Review by sharat87 — CHANGES_REQUESTED** [review]
I've put in comments to a few clarifications and some changes. But it looks quite close to getting merged. Awesome work! 🎉

**Comment by sharat87** [comment]
Also, can we use a smaller, more focused image for the screenshot instead of the whole browser? Something like just:

Screenshot 2021-08-24 at 16 26 33

File: deploy/fat_container/templates/mongo-init.js.sh (hunk 10 above)
Comment (mohanarpit) [reply deploy/fat_container/templates/mongo-init.js.sh:27]: Yes please. Those are redundant now. Can remove.

Comment (sharat87) [reply deploy/fat_container/templates/mongo-init.js.sh:27]: @geekup-legodevops We can remove all the calls to `db.group.insert` and `db.config.insert` in this file.

File: Dockerfile [hunk 17]
```diff
@@ -2,46 +2,40 @@ FROM debian:buster
...
+	/root/.pip \
+	/usr/local/share/doc \
+	/usr/share/doc \
+	/usr/share/man \
+	/var/lib/apt/lists/* \
+	/tmp/*
 
 # Define volumes - Service Layer
-VOLUME [ "/opt/appsmith/data", "/etc/letsencrypt" ]
+VOLUME [ "/opt/appsmith/stacks" ]
```
Comment (sharat87) [inline Dockerfile:38]: This makes for slightly internal-looking path name. Do we require this to be under `/opt/appsmith`? Can we change this to something like `/appsmith-data`? It'll be shorter, and it'll be more memorable when people are writing docker-compose files.

**Review by sharat87 — CHANGES_REQUESTED** [review]
Looks good. Just one more comment.

File: Dockerfile (hunk 17 above)
Comment (geekup-legodevops) [reply Dockerfile:38]: 1/ For the prefix path `/opt/appsmith` , we do not require this to be under this path, we only want to store all resources in the same work directory to keep folder structure clean. We can remove it.
2/ Actually, we have more things in this folder rather than only data. We also have configuration files and SSL certificate under this folder. We suggest that we can make it like `/stacks` or `/appsmith-stacks` to keep it memorable but still match with folder purpose

Comment (sharat87) [reply Dockerfile:38]: Alright, makes sense. Yeah lets go with `/appsmith-stacks`.

**Review by sharat87 — APPROVED** [review]
